machine.js

```diff
@@ -84,8 +84,14 @@ export function state(...args) {
 let invokePromiseType = {
   enter(machine, service, event) {
     this.fn.call(service, service.context, event)
-      .then(data => service.send({ type: 'done', data }))
-      .catch(error => service.send({ type: 'error', error }));
+      .then(data => {
+        if (machine.state.name === service.machine.state.name)
```

Comment:
Since every step generates a new machine, wouldn't it be enough to compare the value of `machine` versus `service.machine`? It would address the case where the state machine has returned to the same state before the promise resolves.